### PR TITLE
K8s

### PR DESCRIPTION
**Pull request by svetomir**
This is large PR which brings two 'big' changes to the beekeeper's design so far.
I could split it in two separate PRs, but I decided not to in order to save time. 
If this is too much for reviewing, please let me know and I will do cherrypicking for 2 PRs.

First 'big' change is related to the **bee** pkg:

- _cluster_ is rewritten so it has more options now
- _nodegroups_ (subset of cluster) are introduced as a concept to allow better management of a cluster and similar bee nodes, they should give more flexibility when creating dynamic clusters
- earlier, cluster use to be []Node, and now, cluster is map[name]NodeGroup, while NodeGroup is map[name]Node
- all this, changes logic how particular Nodes are accessed in the code (checks for example), hence all the checks had to be updated to use new types

Second 'big' change is adding the **k8s** pkg, which has several sub-packages:

- _k8s/bee_ pkg is our implementation of k8s that suits Bee needs, it is a wrapper around other _k8s/*_ pkgs
- other _k8s/*_ pkgs are pure k8s used for CRUD over objects in the k8s cluster

Also, 2 new commands are added:
- start node
- start cluster

### CODE REVIEW

**Review by acud — COMMENTED** [review]
@svetomir kudos for the tons of work that was put into this PR and topic in general. It seems accessible, well thought through and something we could maintain in the long run.

I did observe some repeating patterns which could be improved. They are generally minor but are apparent in the PR, but also in general in `beekeeper` codebase, and we could probably improve on them slowly but surely.
Namely, those are:
- contexts could be propagated from the cobra command to the actual checks (`cmd.Context()` instead of background context in commands)
- I would personally remove the type aliases in the gist of `type SomeCollection map[string]SomeOtherType`. I feel that in general this lacks clarity of which actual types are being passed around and so I personally find passing around a `map[string]SomeType` as clearer. Also, if the type alias does not provide any OTC functionality that needs this type as a receiver then I would personally see less benefit in having it there
- Some functions return an instance instead of a pointer, example is in the configuration functions. I personally would return a pointer there, since I feel that configuration is something the needs to be passed around with potential mutability in mind
- some stream messages have `2` suffix (not sure why)
- named returns - i would personally avoid them. we also converged on this in the `bee` project that we were abusing them too much. they make some of the flow of the code less readable since you need to look specifically for the actual assignments. i would use named returns only when the actual semantics of the values you're returning is not clear and some labeling is needed (but even then, i am not sure i would use blank `return` statements) or when there is a performance optimization consideration to using them
- some receiver methods use value semantics but i think they should use pointer semantics

Also, there seems to be a lot of repetition with `toK8S` methods, which sometimes require type aliases, and I wonder if we could put something together that does this with reflection, since those type aliases are confusing (sometimes they refer to a slice, sometimes a map), and this makes their usage in general a bit more cumbersome (since you don't necessarily know which underlying plumbing you need to access).

Otherwise this looks good!

File: cmd/beekeeper/cmd/create_namespace.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		Long:  `Create Kubernetes namespace.`,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			kubeconfig := c.config.GetString(optionNameCreateConfig)
+			ns := c.config.GetString(optionNameCreateNamespace)
+
+			k := k8s.NewClient(&k8s.ClientOptions{
+				KubeconfigPath: kubeconfig,
+			})
+
+			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/create_namespace.go]: Good point. Updated it.

File: cmd/beekeeper/cmd/print_depths.go [hunk 2]
```diff
@@ -14,30 +14,37 @@ func (c *command) initPrintDepths() *cobra.Command {
...
+			cluster.AddNodeGroup("nodes", ngOptions)
+			ng := cluster.NodeGroup("nodes")
+
+			for i := 0; i < c.config.GetInt(optionNameNodeCount); i++ {
+				if err := ng.AddNode(fmt.Sprintf("bee-%d", i)); err != nil {
+					return fmt.Errorf("adding node bee-%d: %s", i, err)
+				}
 			}
 
 			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/print_depths.go:39]: Good point. Updated it.

File: pkg/k8s/containers/lifecycle.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import v1 "k8s.io/api/core/v1"
+
+// Lifecycle represents Kubernetes Lifecycle
+type Lifecycle struct {
+	PostStart *Handler
+	PreStop   *Handler
+}
+
+// toK8S converts Lifecycle to Kuberntes client object
+func (l Lifecycle) toK8S() *v1.Lifecycle {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/containers/lifecycle.go]: Good point. Done.

File: pkg/k8s/containers/probe.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+// Probe represents Kubernetes Probe
+type Probe struct {
+	Exec      *ExecProbe
+	HTTPGet   *HTTPGetProbe
+	TCPSocket *TCPSocketProbe
+}
+
+// toK8S converts Containers to Kuberntes client object
+func (p Probe) toK8S() *v1.Probe {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/containers/probe.go]: Good point. Done.

File: pkg/k8s/ingress/tls.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package ingress
+
+import ev1b1 "k8s.io/api/extensions/v1beta1"
+
+// TLSs represents service's IngressTLSs
+type TLSs []TLS
+
+// toK8S converts TLSs to Kuberntes client objects
+func (ts TLSs) toK8S() (l []ev1b1.IngressTLS) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/ingress/tls.go]: Replied in the comment bellow.

File: pkg/k8s/ingress/client.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+// Options holds optional parameters for the Client.
+type Options struct {
+	Annotations map[string]string
+	Labels      map[string]string
+	Spec        Spec
+}
+
+// Set creates Ingress, if Ingress already exists does nothing
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/ingress/client.go]: Replied in the comment bellow.

File: pkg/k8s/persistentvolumeclaim/client.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+// Options holds optional parameters for the Client.
+type Options struct {
+	Annotations map[string]string
+	Labels      map[string]string
+	Spec        PersistentVolumeClaimSpec
+}
+
+// Set creates PersistentVolumeClaim, if PersistentVolumeClaim already exists updates in place
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/persistentvolumeclaim/client.go]: Replied in the comment bellow.

File: cmd/beekeeper/cmd/helpers.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"fmt"
+
+	"github.com/ethersphere/beekeeper/pkg/bee"
+	k8sBee "github.com/ethersphere/beekeeper/pkg/k8s/bee"
+)
+
+// newBeeDefaultConfig returns default Bee node configuration
+func newBeeDefaultConfig() k8sBee.Config {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/helpers.go:11]: Good point. I updated it.

File: cmd/beekeeper/cmd/helpers.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		TracingEnabled:       false,
+		TracingEndpoint:      "jaeger-operator-jaeger-agent.observability:6831",
+		TracingServiceName:   "bee",
+		Verbosity:            5,
+		WelcomeMessage:       "Welcome to the Swarm, you are Bee-ing connected!",
+	}
+}
+
+// newDefaultNodeGroupOptions returns default node group options
+func newDefaultNodeGroupOptions() bee.NodeGroupOptions {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/helpers.go:48]: Good point. I updated it.

File: cmd/beekeeper/cmd/print_overlay.go [hunk 10]
```diff
@@ -14,30 +14,36 @@ func (c *command) initPrintOverlay() *cobra.Command {
...
+			cluster.AddNodeGroup("nodes", ngOptions)
+			ng := cluster.NodeGroup("nodes")
+
+			for i := 0; i < c.config.GetInt(optionNameNodeCount); i++ {
+				if err := ng.AddNode(fmt.Sprintf("bee-%d", i)); err != nil {
+					return fmt.Errorf("adding node bee-%d: %s", i, err)
+				}
 			}
 
 			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/print_overlay.go:39]: Good point. Updated it.

File: cmd/beekeeper/cmd/print_topologies.go [hunk 11]
```diff
@@ -14,36 +14,42 @@ func (c *command) initPrintTopologies() *cobra.Command {
...
+			cluster.AddNodeGroup("nodes", ngOptions)
+			ng := cluster.NodeGroup("nodes")
+
+			for i := 0; i < c.config.GetInt(optionNameNodeCount); i++ {
+				if err := ng.AddNode(fmt.Sprintf("bee-%d", i)); err != nil {
+					return fmt.Errorf("adding node bee-%d: %s", i, err)
+				}
 			}
 
 			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/print_topologies.go:39]: Good point. Updated it.

File: cmd/beekeeper/cmd/start_cluster.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+		bootnodeCount int
+		nodeCount     int
+	)
+
+	cmd := &cobra.Command{
+		Use:   "cluster",
+		Short: "Start Bee cluster",
+		Long:  `Start Bee cluster.`,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/start_cluster.go]: Good point. Updated it.

File: cmd/beekeeper/cmd/print_peers.go [hunk 13]
```diff
@@ -14,31 +14,37 @@ func (c *command) initPrintPeers() *cobra.Command {
...
+			cluster.AddNodeGroup("nodes", ngOptions)
+			ng := cluster.NodeGroup("nodes")
+
+			for i := 0; i < c.config.GetInt(optionNameNodeCount); i++ {
+				if err := ng.AddNode(fmt.Sprintf("bee-%d", i)); err != nil {
+					return fmt.Errorf("adding node bee-%d: %s", i, err)
+				}
 			}
 
 			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/print_peers.go]: Good point. Updated it.

File: cmd/beekeeper/cmd/start_node.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+		nodeName         string
+		standalone       bool
+	)
+
+	cmd := &cobra.Command{
+		Use:   "node",
+		Short: "Start Bee node",
+		Long:  `Start Bee node.`,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/start_node.go]: Good point. Updated it.

File: pkg/bee/cluster.go [hunk 15]
```diff
@@ -5,515 +5,293 @@ import (
...
-		msgs = append(msgs, m)
-	}
+// ClusterOverlays represents overlay addresses of all nodes in the cluster
+type ClusterOverlays map[string]NodeGroupOverlays
 
-	sort.SliceStable(msgs, func(i, j int) bool {
-		return msgs[i].Index < msgs[j].Index
-	})
+// Overlays returns ClusterOverlays
+func (c *Cluster) Overlays(ctx context.Context) (overlays ClusterOverlays, err error) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/cluster.go:164]: Answered bellow

File: pkg/bee/cluster.go [hunk 16]
```diff
@@ -5,515 +5,293 @@ import (
...
 }
 
-// OverlaysStreamMsg represents message sent over the OverlaysStream channel
-type OverlaysStreamMsg struct {
-	Address swarm.Address
-	Index   int
-	Error   error
-}
+// ClusterPeers represents peers of all nodes in the cluster
+type ClusterPeers map[string]NodeGroupPeers
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/cluster.go:180]: Answered bellow

File: pkg/bee/cluster.go [hunk 17]
```diff
@@ -5,515 +5,293 @@ import (
...
 }
 
-// Peers returns ordered list of peers of all nodes in the cluster
-func (c *Cluster) Peers(ctx context.Context) (peers [][]swarm.Address, err error) {
-	var msgs []PeersStreamMsg
-	for m := range c.PeersStream(ctx) {
-		msgs = append(msgs, m)
-	}
+// ClusterSettlements represents settlements of all nodes in the cluster
+type ClusterSettlements map[string]NodeGroupSettlements
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/cluster.go:199]: Answered bellow

File: pkg/bee/cluster.go [hunk 18]
```diff
@@ -5,515 +5,293 @@ import (
...
 	return
 }
 
-// SentReceived object
-type SentReceived struct {
-	Received int
-	Sent     int
-}
+// ClusterTopologies represents Kademlia topology of all nodes in the cluster
+type ClusterTopologies map[string]NodeGroupTopologies
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/cluster.go:226]: Answered bellow

File: pkg/bee/node.go [hunk 19]
```diff
@@ -17,48 +17,56 @@ import (
...
 	DebugAPIURL         *url.URL
 	DebugAPIInsecureTLS bool
 }
 
-// NewNode returns new node
-func NewNode(opts NodeOptions) Node {
-	return Node{
-		api: api.NewClient(opts.APIURL, &api.ClientOptions{HTTPClient: &http.Client{Transport: &http.Transport{
+// NewClient returns Bee client
+func NewClient(opts ClientOptions) (c Client) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/node.go]: Updated it.

File: pkg/bee/nodegroup.go [hunk 20]
```diff
@@ -0,0 +1,593 @@
...
+		DebugAPIURL:         dURL,
+		DebugAPIInsecureTLS: g.cluster.debugAPIInsecureTLS,
+	})
+	g.nodes[name] = &c
+
+	return
+}
+
+// NodeGroupAddresses represents addresses of all nodes in the node group
+type NodeGroupAddresses map[string]Addresses
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/nodegroup.go:79]: It is :)

File: pkg/bee/nodegroup.go [hunk 21]
```diff
@@ -0,0 +1,593 @@
...
+			return nil, fmt.Errorf("%s: %w", m.Name, m.Error)
+		}
+		addrs[m.Name] = m.Addresses
+	}
+
+	return
+}
+
+// AddressesStreamMsg2 represents message sent over the AddressStream channel
+type AddressesStreamMsg2 struct {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/bee/nodegroup.go]: Good find. :) I overlooked it. I had 2 clients at one moment (legacy one, too).

File: pkg/check/balances/balances.go [hunk 22]
```diff
@@ -23,18 +24,19 @@ type Options struct {
 }
 
 // Check executes balances check
-func Check(c bee.Cluster, o Options, pusher *push.Pusher, pushMetrics bool) (err error) {
+func Check(c *bee.Cluster, o Options, pusher *push.Pusher, pushMetrics bool) (err error) {
 	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/check/balances/balances.go:28]: Done.

File: pkg/k8s/bee/helpers.go [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+type setInitContainersOptions struct {
+	ClefEnabled         bool
+	ClefImage           string
+	ClefImagePullPolicy string
+	ClefPassword        string
+	LibP2PEnabled       bool
+	SwarmEnabled        bool
+}
+
+func setInitContainers(o setInitContainersOptions) (inits containers.Containers) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/bee/helpers.go:23]: Answered bellow.

File: pkg/k8s/bee/node.go [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+	"github.com/ethersphere/beekeeper/pkg/k8s/service"
+	"github.com/ethersphere/beekeeper/pkg/k8s/serviceaccount"
+	"github.com/ethersphere/beekeeper/pkg/k8s/statefulset"
+)
+
+const (
+	portHTTP = 80
+)
+
+// NodeStartOptions ...
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/bee/node.go:22]: Added a proper comment.

File: pkg/k8s/configmap/configmap.go [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+type Options struct {
+	Annotations map[string]string
+	Labels      map[string]string
+	Immutable   bool
+	Data        map[string]string
+	BinaryData  map[string][]byte
+}
+
+// Set creates ConfigMap, if ConfigMap already exists updates in place
+func (c Client) Set(ctx context.Context, name, namespace string, o Options) (err error) {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/configmap/configmap.go]: Answered bellow

File: pkg/k8s/containers/containers.go [hunk 26]
```diff
@@ -0,0 +1,73 @@
+package containers
+
+import (
+	v1 "k8s.io/api/core/v1"
+)
+
+// Containers represents Kubernetes Containers
+type Containers []Container
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/containers/containers.go:8]: Answered bellow

File: pkg/k8s/containers/env.go [hunk 27]
```diff
@@ -0,0 +1,168 @@
+package containers
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// EnvVars represents Kubernetes EnvVars
+type EnvVars []EnvVar
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/containers/env.go:9]: Answered bellow

File: pkg/k8s/containers/handler.go [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+
+// Handler represents Kubernetes Handler
+type Handler struct {
+	Exec      *ExecHandler
+	HTTPGet   *HTTPGetHandler
+	TCPSocket *TCPSocketHandler
+}
+
+// toK8S converts Handler to Kuberntes client object
+func (h Handler) toK8S() v1.Handler {
```
[the remark this answers is not in the document]
Comment (svetomir) [reply pkg/k8s/containers/handler.go]: Updated it.

File: cmd/beekeeper/cmd/start_cluster.go [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+				"app.kubernetes.io/part-of":   ngName,
+				"app.kubernetes.io/version":   strings.Split(image, ":")[1],
+			}
+			cluster.AddNodeGroup(ngName, ngOptions)
+			ng := cluster.NodeGroup(ngName)
+
+			nConfig := newBeeDefaultConfig()
+			nConfig.Bootnodes = setupBootnodesDNS(bootnodeCount, c.config.GetString(optionNameStartNamespace))
+			for i := 0; i < nodeCount; i++ {
+				if err := ng.StartNode(ctx, bee.StartNodeOptions{
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/start_cluster.go]: Definitely, I will add concurrency, but once I optimise k8s pkg for it. That is next step once this PR is merged.

File: cmd/beekeeper/cmd/start_node.go [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+
+			cluster := bee.NewCluster(clusterName, bee.ClusterOptions{
```
[the remark this answers is not in the document]
Comment (svetomir) [reply cmd/beekeeper/cmd/start_node.go:38]: This doesn't seem like a blocker now. I left this when I finish dynamic cluster feature, as I will have more insight on everything needed for it. I planned to refactor this once I deal with priorites.

**Comment by svetomir** [comment]
Hey, @acud thanks a lot for the review.  Comments and advices were very detailed and I am grateful for it. I am glad you find it maintainable.
We definitely should improve beekeeper codebase in the long run as now it becomes apparent in which direction and how it should be developed.
I already tackled some of those based on your review. Here are my comments about it:
* contexts could be propagated from the cobra command to the actual checks (cmd.Context() instead of background context in commands)
    * Good point. I updated this.
* I would personally remove the type aliases in the gist of type SomeCollection map[string]SomeOtherType. I feel that in general this lacks clarity of which actual types are being passed around and so I personally find passing around a map[string]SomeType as clearer. Also, if the type alias does not provide any OTC functionality that needs this type as a receiver then I would personally see less benefit in having it there
    * map[string]SomeType, etc. are regular types (not type aliases) needed to mimic k8s types and allow easier type conversion. I used them for clarity because they are mostly part of some more complex type (struct) and this complexity comes with big number of k8s types.
* Some functions return an instance instead of a pointer, example is in the configuration functions. I personally would return a pointer there, since I feel that configuration is something the needs to be passed around with potential mutability in mind
    * Good point. I updated this.
* some stream messages have 2 suffix (not sure why)
    * Good find. It was a legacy from an old client and I overlooked it. Anyways, I renamed it. Thanks :)
* named returns - i would personally avoid them. we also converged on this in the bee project that we were abusing them too much. they make some of the flow of the code less readable since you need to look specifically for the actual assignments. i would use named returns only when the actual semantics of the values you're returning is not clear and some labelling is needed (but even then, i am not sure i would use blank return statements) or when there is a performance optimisation consideration to using them
    * Whole beekeeper codebase is written in this manner and I would keep it as it is. I find it more readable. Anyways, we can review this again with another PR that will be focused on this.
* some receiver methods use value semantics but i think they should use pointer semantics
    * Good point. I updated this.
* Also, there seems to be a lot of repetition with toK8S methods, which sometimes require type aliases, and I wonder if we could put something together that does this with reflection, since those type aliases are confusing (sometimes they refer to a slice, sometimes a map), and this makes their usage in general a bit more cumbersome (since you don't necessarily know which underlying plumbing you need to access).
    * This way we have explicit methods for conversion to k8s types. Also, in the long run we’ll have less troubles replacing them with generics once they become available in Go. :)

**Review by janos — APPROVED** [review]
A lot of great work is here @svetomir. Thank you! My review is a bit subjective, so it is good that @acud gave it a very good one.

**Review by acud — APPROVED** [review]
Thanks for addressing my comments @svetomir. Looks good and thanks again for all the work done here!

**Pull request merged**